### PR TITLE
relocate.py: parallelize test replacement logic

### PR DESCRIPTION
**Pull request by nhanford**
Parallelizing functions and increasing prefix replacement performance in the relocate.py module.
Also changed signatures and calls.

### CODE REVIEW

File: lib/spack/spack/directory_layout.py [hunk 1]
```diff
@@ -214,6 +214,7 @@ def __init__(self, root, **kwargs):
         self.metadata_dir        = '.spack'
         self.deprecated_dir      = 'deprecated'
         self.spec_file_name      = 'spec.yaml'
+        self.deploy_file_name    = 'deploy.yaml'
```
Comment (becker33) [inline lib/spack/spack/directory_layout.py]: Is this accidentally included from another PR?

File: spack.code-workspace [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+	"folders": [
+		{
+			"path": "."
+		}
+	]
+}
```
Comment (becker33) [inline spack.code-workspace]: I think this shouldn't be included in the PR.

**Review by becker33 — CHANGES_REQUESTED** [review]
Two stray commits in addition to needing a new sbang test

**Comment by eugeneswalker** [comment]
I realize the point of this PR is to parallelize test replacement logic, but I thought it would make sense to note here that `sbang` relocation does not work in the particular case that `install_tree: padded_length:` is set in conjunction with a non-default `install_tree: root:` as of 03f6717c3883b12705a3f3598eb49ce02cecb16c from `2021-01-24`.

This problem is described in Issue https://github.com/spack/spack/issues/20766
One solution to the problem is in PR https://github.com/spack/spack/pull/20768

I'm really, really hoping we can address the sbang issue in some form for the next tag release of Spack, such as `0.16.1`. Whether that is done using my PR, or a better solution proposed by others.

**Review by becker33 — APPROVED** [review]

**Pull request merged**